imports/client/components/LoginForm.tsx

```diff
@@ -1,14 +1,48 @@
-import React, { useCallback, useState } from "react";
-import AccountForm, { AccountFormFormat } from "./AccountForm";
+import React, { useCallback, useEffect, useState } from "react";
+import { useLocation } from "react-router-dom";
+import AccountForm, {
+  AccountFormFormat,
+  InvitationEnrollmentMode,
+} from "./AccountForm";
 
 type LoginFormFormat =
   | AccountFormFormat.LOGIN
-  | AccountFormFormat.REQUEST_PW_RESET;
+  | AccountFormFormat.REQUEST_PW_RESET
+  | AccountFormFormat.INVITATION_WELCOME
+  | AccountFormFormat.ENROLL;
 
 const LoginForm = () => {
-  const [format, setFormat] = useState<LoginFormFormat>(
-    AccountFormFormat.LOGIN,
-  );
+  const location = useLocation();
+
+  const [format, setFormat] = useState<LoginFormFormat | undefined>(undefined);
+
+  const [huntInvitationCode, setHuntInvitationCode] = useState<string>("");
+
+  useEffect(() => {
+    if (!format) {
+      if (location.state?.pathname) {
+        const invitationPattern = /^\/join\/(.+)$/;
+        const matchResult = invitationPattern.exec(location.state.pathname);
+        if (matchResult?.[1]) {
+          setHuntInvitationCode(matchResult[1]);
+          if (format == null) {
+            setFormat(AccountFormFormat.INVITATION_WELCOME);
+            return;
+          }
+        }
+      }
+      setFormat(AccountFormFormat.LOGIN);
```

Comment:
When we navigate to the login form in `AuthenticatedPage`, the value that we set for state is unconstrained (react-router just types it as `any`). We can make this cleaner by just putting more structure into state, right?

Reply:
If I understand correctly - you're suggesting that in `AuthenticatedPage`, rather than just setting `state={{ pathname, search }}`, we could do something more structured such that here, we could access that more structured data instead of having to extract the invitation code out from the pathname using a regex like this.

But I don't see what AuthenticatedPage has access to that would make that any easier - all it knows is the raw location, from which it pulls the pathname / search. So it seems to me that it would have to do exactly the same sort of path extraction in order to stash it in a more structured form, and if you're going to have to do that somewhere, it seems better to do it where it is needed rather than in a central place.

Ideally we could find a way to apply the path matching from Route.tsx directly to the path so we could actually extract something like a page identifier and the invitation code parameter, but I wasn't seeing an obvious way to do that outside of the standard routing mechanism.

Let me know if I misunderstood.

Reply:
Oh yeah, that's right.

Hmm. What about not using the authentication helpers on the join route, and having the join route itself redirect to the login/enroll page?

Reply:
Done. I feel a little uneasy from a security perspective about allowing a page to manage authentication itself, but I suppose its justified in this case.